src/librustc_typeck/check/demand.rs

```diff
@@ -261,6 +261,28 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (_, &ty::TyRef(_, checked)) => {
+                if self.infcx.can_sub(self.param_env, checked.ty, &expected).is_ok() {
+                    match expr.node {
+                        hir::ExprAddrOf(_, ref expr) => {
+                            if let Ok(code) = self.tcx.sess.codemap().span_to_snippet(expr.span) {
+                                return Some(format!("try with `{}`", code));
```

Comment:
I guess @GuillaumeGomez that this particular macro is triggering this case. Try having the macro produce `{&x}` for example -- although this is *also* an interesting data point.